### PR TITLE
[10][IMP] sale_exception add field

### PR DESCRIPTION
**Pull request by hparfr**
Add a reverse field to leverage optimizations of base_exception.

Since base_exception@10.0.2.0.0, if there is a relation from exception.rule to a model, it will be used to write results once per rule instead of iterate through all the sale order records.

### CODE REVIEW

**Review by PierrickBrun — APPROVED** [review]

**Review by chafique-delli — APPROVED** [review]
Thank you @hparfr  for this improvement

**Comment by pedrobaeza** [comment]
I think this is not going to be very useful when a bit of time has passed and you have tons of past orders in exception.

**Comment by pedrobaeza** [comment]
And this will also slow down a lot the load of the exception form. This should go as an smart-button that navigates to list.

**Comment by hparfr** [comment]
>I think this is not going to be very useful when a bit of time has passed and you have tons of past orders in exception.

The bigger len(sale.order)  / len(exception.rule) is, the bigger this commit will improve perfs.

>And this will also slow down a lot the load of the exception form. This should go as an smart-button that navigates to list.

I don't get it. Which form ? This one  https://github.com/OCA/server-tools/blob/10.0/base_exception/views/base_exception_view.xml#L18 ?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
See me previous comment.

**Comment by hparfr** [comment]
Added test that can pass only after the following PR is merged in server-tools https://github.com/OCA/server-tools/pull/1480

It's not necessary anymore to scan all the sales to get accurate results.
It's useful when lot of sales can't be excluded by domain (like state=draft).

File: sale_exception/__manifest__.py [hunk 1]
```diff
@@ -4,14 +4,15 @@
 
 {'name': 'Sale Exception',
  'summary': 'Custom exceptions on sale order',
- 'version': '10.0.2.0.0',
+ 'version': '10.0.2.0.1',
  'category': 'Generic Modules/Sale',
  'author': "Akretion, Sodexis, Camptocamp, Odoo Community Association (OCA)",
  'website': 'http://www.akretion.com',
```
Comment (nikul-serpentcs) [inline sale_exception/__manifest__.py]: URL Should be `https://github.com/OCA/sale-workflow`

File: sale_exception/tests/__init__.py [hunk 2]
```diff
@@ -3,3 +3,4 @@
 # License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
 
 from . import test_sale_exception
+from . import test_multi_records
```
Comment (nikul-serpentcs) [inline sale_exception/tests/__init__.py]: @hparfr add a new blank line

**Comment by florian-dacosta** [comment]
This PR depends on https://github.com/OCA/server-tools/pull/1601
It fixes the module after the last refactore.

@pedrobaeza Could you please recheck, I believe your request for change is outdated.

@bealdav @mourad-ehm @sebastienbeau

**Review by florian-dacosta — APPROVED** [review]

**Review by bealdav — APPROVED** [review]
LGTM

**Review by mourad-ehm — APPROVED** [review]

**Comment by florian-dacosta** [comment]
I reverted oca_dependencies since https://github.com/OCA/server-tools/pull/1601 has been merged.

This PR is now ready to be merged.

**Review by sebastienbeau — APPROVED** [review]
LGTM code review

**Comment by sebastienbeau** [comment]
@pedrobaeza it's ok for you?

**Comment by sebastienbeau** [comment]
The same PR in 12 have been merged so let's merge it https://github.com/OCA/sale-workflow/pull/873

**Pull request merged**